crates/red_knot_python_semantic/src/types/narrow.rs

```diff
@@ -51,15 +51,30 @@ fn all_narrowing_constraints_for_pattern<'db>(
     db: &'db dyn Db,
     pattern: PatternConstraint<'db>,
 ) -> NarrowingConstraints<'db> {
-    NarrowingConstraintsBuilder::new(db, Constraint::Pattern(pattern)).finish()
+    NarrowingConstraintsBuilder::new(
+        db,
+        Constraint {
+            node: ConstraintNode::Pattern(pattern),
+            negative: false,
+        },
+    )
+    .finish()
 }
 
 #[salsa::tracked(return_ref)]
 fn all_narrowing_constraints_for_expression<'db>(
     db: &'db dyn Db,
     expression: Expression<'db>,
+    negative: bool,
 ) -> NarrowingConstraints<'db> {
-    NarrowingConstraintsBuilder::new(db, Constraint::Expression(expression)).finish()
+    NarrowingConstraintsBuilder::new(
+        db,
+        Constraint {
+            node: ConstraintNode::Expression(expression),
+            negative,
+        },
```

Comment:
Is it worth re-building a `Constraint` struct here vs just passing in the two elements separately to the builder? It mostly seems like it just results in more verbose usage inside the builder (`self.constraint.negative` vs `self.negative`).